### PR TITLE
👻 Enhance regex patterns

### PR DESCRIPTION
**Pull request by Adirio**
# Descrition
- Used non-capturing groups when the info is not relevant
- Simplified some expressions

# Motivation
While non-capturing groups should be a bit more performant that capturing groups, the improvement will be negligible. The motivations is mainly improving the clarity and readability of the regex expressions by simplifying some patters and not having unused capture groups.

/kind cleanup

### CODE REVIEW

**Comment by Adirio** [comment]
I used 👻 as this is a minor change that I don't think is worth to include in changelogs, but the verifier is not understanding it despite being mentioned in CONTRIBUTING.md. Guess I could change it to ✨ if needed.

File: pkg/internal/validation/dns.go [hunk 1]
```diff
@@ -26,9 +26,9 @@ import (
 // "k8s.io/apimachinery" is needed, re-consider whether to add the dependency.
 
 const (
-	dns1123LabelFmt     string = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"
+	dns1123LabelFmt     string = "[a-z0-9](?:[-a-z0-9]*[a-z0-9])?"
 	dns1123SubdomainFmt string = dns1123LabelFmt + "(\\." + dns1123LabelFmt + ")*"
-	dns1035LabelFmt     string = "[a-z]([-a-z0-9]*[a-z0-9])?"
+	dns1035LabelFmt     string = "[a-z](?:[-a-z0-9]*[a-z0-9])?"
```
Comment (camilamacedo86) [inline pkg/internal/validation/dns.go:31]: @Adirio, 

For we not add ore one dependency to the project we decided to copy and paste the implementation to validate. 
We should not change this file unless it was changed in the `k8s.io/apimachinery` 

See: https://github.com/kubernetes-sigs/kubebuilder/blob/175952919092cd7304e5e8aeddb9cce37b6a614b/pkg/internal/validation/dns.go#L24-L26

Comment (camilamacedo86) [reply pkg/internal/validation/dns.go:31]: Could you please revert the changes here?

File: pkg/plugin/plugin.go [hunk 2]
```diff
@@ -92,7 +92,7 @@ func ParseVersion(version string) (v Version, err error) {
 	}
 
 	// Parse stage suffix, if any.
-	v.Stage = strings.TrimPrefix(submatches[3], "-")
```
Comment (camilamacedo86) [inline pkg/plugin/plugin.go:95]: Why not trim the spaces?

Comment (Adirio) [reply pkg/plugin/plugin.go:95]: We are timming the dash (`-`), not the spaces, and by modifying the regex as I did the dash is not part of the capture group now, so no need to trim it.

File: pkg/plugin/plugin.go [hunk 3]
```diff
@@ -75,15 +75,15 @@ func ParseVersion(version string) (v Version, err error) {
 		return v, errors.New("plugin version is empty")
 	}
 
-	// A valid version string will have 4 submatches, each of which may be empty: the full string, "v",
-	// the integer, and the stage suffix. Invalid version strings do not have 4 submatches.
+	// A valid version string will have 3 submatches, each of which may be empty: the full string,
+	// the integer, and the stage suffix. Invalid version strings do not have 3 submatches.
 	submatches := verRe.FindStringSubmatch(version)
-	if len(submatches) != 4 {
+	if len(submatches) != 3 {
```
Comment (camilamacedo86) [inline pkg/plugin/plugin.go:81]: Why the changes here would make it easier to understand?

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
IMO the change in https://github.com/kubernetes-sigs/kubebuilder/pull/1745/files#r512653755 needs to be revert. I am not sure why the changes made make the code more understandable or not. I think it also needs to be checked by @estroz.

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (Adirio) [reply pkg/plugin/plugin.go:81]: One of the submatches was `v`. We are not using `v` for anything so getting a submatch for it is inneficient. I just removed that submatch, and therefore I had to subtract 1 to the total length and the indexes.

File: pkg/internal/validation/dns.go (hunk 1 above)
Comment (Adirio) [reply pkg/internal/validation/dns.go:31]: I would agree with you if the code was copy-pasted, but it has been already modified. You can [check it](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go). So Convering that capture group into a non-capture group tells the future us reading the code that we are not using that capture group to take a substring out of the regex.

Comment (camilamacedo86) [reply pkg/internal/validation/dns.go:31]: Could we keep the sample implementation of https://github.com/kubernetes/apimachinery/blob/v0.18.6/pkg/util/validation/validation.go ? (copy and paste)

Comment (camilamacedo86) [reply pkg/internal/validation/dns.go:31]: Catcher. I would prefer if possible we keep a copy and paste of https://github.com/kubernetes/apimachinery/blob/v0.18.6/pkg/util/validation/validation.go whcih would allow us to move to use the dependency if we wish without breaking changes. However, I will let it up to you and @ezric

**Review by camilamacedo86 — COMMENTED** [review]
HI @Adirio, 

I am ok with the changes, however, I think it needs the @estroz ok as well if possible. 
My concern would only : https://github.com/kubernetes-sigs/kubebuilder/pull/1745#discussion_r512703138

Also, note that it is no passing in the CI. could you please fix that?

**Review by camilamacedo86 — DISMISSED** [review]

File: pkg/model/resource/options_test.go [hunk 4]
```diff
@@ -47,27 +47,27 @@ var _ = Describe("Resource Options", func() {
 			options := &Options{Group: "crew", Version: "1", Kind: "FirstMate"}
 			Expect(options.Validate()).NotTo(Succeed())
 			Expect(options.Validate().Error()).To(ContainSubstring(
-				`version must match ^v\d+(alpha\d+|beta\d+)?$ (was 1)`))
+				`version must match ^v\d+(?:alpha\d+|beta\d+)?$ (was 1)`))
```
Comment (camilamacedo86) [inline pkg/model/resource/options_test.go:50]: IMO : it should be a const

Comment (Adirio) [reply pkg/model/resource/options_test.go:50]: I also thought of that to prevent having it hardcoded a doxen times the thing is that the string is not exported from its package. I think the better approach would be to create an expecific error instead a generic one and test if this yields this kind of error. But I just wanted to do the less changes to get it to pass Travis for now.

**Review by estroz — APPROVED** [review]
/lgtm
